container.go

```diff
@@ -0,0 +1,342 @@
+	title := "HCSShim::OpenContainer"
+	var (
+		handle  HcsSystem
+		resultp *uint16
+	)
+
+	if err := hcsOpenComputeSystem(id, &handle, &resultp); err != nil {
+		err = makeErrorf(err, title, "id=%s", id)
+		logrus.Error(err)
+		return nil, err
+	}
+	// result not currently used, but needs to be properly freed.
+	if resultp != nil {
+		result := convertAndFreeCoTaskMemString(resultp)
+		logrus.Debugf("%s returned result: %s", title, result)
+	}
+
+	container := &hcsShimContainer{
+		handle: handle,
+	}
+
+	return container, nil
+}
+
+// Start synchronously starts the container.
+func (container *hcsShimContainer) Start() error {
+	title := "HCSShim::Container::Start"
+	var (
+		options string
+		resultp *uint16
+	)
+
+	if err := hcsStartComputeSystem(container.handle, options, &resultp); err != nil {
+		err = makeError(err, title, "")
+		logrus.Error(err)
+		return err
+	}
+	// result not currently used, but needs to be properly freed.
+	if resultp != nil {
+		result := convertAndFreeCoTaskMemString(resultp)
+		logrus.Debugf("%s returned result: %s", title, result)
+	}
+
+	return nil
+}
+
+// Shutdown requests a container shutdown, but it may not actually be shutdown until Wait() succeeds.
+func (container *hcsShimContainer) Shutdown() error {
+	title := "HCSShim::Container::Shutdown"
+	var (
+		options string
+		resultp *uint16
+	)
+
+	if err := hcsShutdownComputeSystem(container.handle, options, &resultp); err != nil {
+		err = makeError(err, title, "")
+		logrus.Error(err)
+		return err
+	}
+	// result not currently used, but needs to be properly freed.
+	if resultp != nil {
+		result := convertAndFreeCoTaskMemString(resultp)
+		logrus.Debugf("%s returned result: %s", title, result)
+	}
+
+	return nil
+}
+
+// Terminate requests a container terminate, but it may not actually be terminated until Wait() succeeds.
+func (container *hcsShimContainer) Terminate() error {
+	title := "HCSShim::Container::Terminate"
+	var (
+		options string
+		resultp *uint16
+	)
+
+	if err := hcsTerminateComputeSystem(container.handle, options, &resultp); err != nil {
+		err = makeError(err, title, "")
+		logrus.Error(err)
+		return err
+	}
+	// result not currently used, but needs to be properly freed.
+	if resultp != nil {
+		result := convertAndFreeCoTaskMemString(resultp)
+		logrus.Debugf("%s returned result: %s", title, result)
+	}
+
+	return nil
+}
+
+// Wait synchronously waits for the container to shutdown or terminate.
+func (container *hcsShimContainer) Wait() error {
+	_, err := container.waitTimeoutInternal(syscall.INFINITE)
+	return err
+}
+
+// WaitTimeout synchronously waits for the container to terminate or the duration to elapse. It
+// returns true if the event occurs successfully without timeout.
+func (container *hcsShimContainer) WaitTimeout(timeout time.Duration) (bool, error) {
+	// TODO can this overflow uint32
+	return container.waitTimeoutInternal(uint32(timeout.Nanoseconds() / (1000 * 1000)))
```

Comment:
time.Duration doesn't have a millisecond option. I'm using nano and dividing by ms/ns to get ms. I'll find a way to make that more clear in the code.